modules/core/src/opencl/intel_gemm.cl

```diff
@@ -0,0 +1,1081 @@
+    float4 dot01 = (start_index != 0) ? vload4(0, dst_write0 + 1 * ldC) : (float)beta * vload4(0, dst_write0 + 1 * ldC);
+    float4 dot02 = (start_index != 0) ? vload4(0, dst_write0 + 2 * ldC) : (float)beta * vload4(0, dst_write0 + 2 * ldC);
+    float4 dot03 = (start_index != 0) ? vload4(0, dst_write0 + 3 * ldC) : (float)beta * vload4(0, dst_write0 + 3 * ldC);
+    float4 dot04 = (start_index != 0) ? vload4(0, dst_write0 + 4 * ldC) : (float)beta * vload4(0, dst_write0 + 4 * ldC);
+    float4 dot05 = (start_index != 0) ? vload4(0, dst_write0 + 5 * ldC) : (float)beta * vload4(0, dst_write0 + 5 * ldC);
+    float4 dot06 = (start_index != 0) ? vload4(0, dst_write0 + 6 * ldC) : (float)beta * vload4(0, dst_write0 + 6 * ldC);
+    float4 dot07 = (start_index != 0) ? vload4(0, dst_write0 + 7 * ldC) : (float)beta * vload4(0, dst_write0 + 7 * ldC);
+   
+    int end_index = min(start_index + stride, K);
+    int w = start_index;
+    while( w + TILE_K <= end_index ) {
+        arow0 = (float)alpha * vload2(0, src0_read + 0 * ldA);
+        arow1 = (float)alpha * vload2(0, src0_read + 1 * ldA);
+        arow2 = (float)alpha * vload2(0, src0_read + 2 * ldA);
+        arow3 = (float)alpha * vload2(0, src0_read + 3 * ldA);
+        arow4 = (float)alpha * vload2(0, src0_read + 4 * ldA);
+        arow5 = (float)alpha * vload2(0, src0_read + 5 * ldA);
+        arow6 = (float)alpha * vload2(0, src0_read + 6 * ldA);
+        arow7 = (float)alpha * vload2(0, src0_read + 7 * ldA);
+
+#define MM_DOT_PRODUCT(index, suffix)   \
+        brow = vload4(0, src1_read0);  src1_read0 += ldB; \
+        dot00 = mad((float4)(intel_sub_group_shuffle(arow0.s##suffix,index)),brow,dot00); \
+        dot01 = mad((float4)(intel_sub_group_shuffle(arow1.s##suffix,index)),brow,dot01); \
+        dot02 = mad((float4)(intel_sub_group_shuffle(arow2.s##suffix,index)),brow,dot02); \
+        dot03 = mad((float4)(intel_sub_group_shuffle(arow3.s##suffix,index)),brow,dot03); \
+        dot04 = mad((float4)(intel_sub_group_shuffle(arow4.s##suffix,index)),brow,dot04); \
+        dot05 = mad((float4)(intel_sub_group_shuffle(arow5.s##suffix,index)),brow,dot05); \
+        dot06 = mad((float4)(intel_sub_group_shuffle(arow6.s##suffix,index)),brow,dot06); \
+        dot07 = mad((float4)(intel_sub_group_shuffle(arow7.s##suffix,index)),brow,dot07);
+
+        MM_DOT_PRODUCT(0,0);
+        MM_DOT_PRODUCT(0,1);
+        MM_DOT_PRODUCT(1,0);
+        MM_DOT_PRODUCT(1,1);
+        MM_DOT_PRODUCT(2,0);
+        MM_DOT_PRODUCT(2,1);
+        MM_DOT_PRODUCT(3,0);
+        MM_DOT_PRODUCT(3,1);
+        MM_DOT_PRODUCT(4,0);
+        MM_DOT_PRODUCT(4,1);
+        MM_DOT_PRODUCT(5,0);
+        MM_DOT_PRODUCT(5,1);
+        MM_DOT_PRODUCT(6,0);
+        MM_DOT_PRODUCT(6,1);
+        MM_DOT_PRODUCT(7,0);
+        MM_DOT_PRODUCT(7,1);
+#undef MM_DOT_PRODUCT
+
+        src0_read += TILE_K;
+        w += TILE_K;
+    }
+
+    vstore4(dot00, 0, dst_write0); dst_write0 += ldC;
```

Comment:
We can't use vectorized store in case of non-aligned data/sizes.
For example, for contiguous matrix 3x3 this will garbage memory in next row.

We need to add more checks into host code.

P.S. Similar problem on `vload` statements.